typing/typetexp.ml

```diff
@@ -50,8 +51,209 @@ type error =
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
-(** Map indexed by type variable names. *)
-module TyVarMap = Misc.Stdlib.String.Map
+module TyVarEnv : sig
+  val reset : unit -> unit
+  (* see mli file *)
+
+  val is_in_scope : string -> bool
+
+  val add : string -> type_expr -> unit
+  (* add a global type variable to the environment *)
+
+  val with_local_scope : (unit -> 'a) -> 'a
+  (* see mli file *)
+
+  type poly_univars
+  val with_univars : poly_univars -> (unit -> 'a) -> 'a
+  (* evaluate with a locally extended set of univars *)
+
+  val make_poly_univars : string list -> poly_univars
+  (* see mli file *)
+
+  val check_poly_univars : Env.t -> Location.t -> poly_univars -> type_expr list
+  (* see mli file *)
+
+  type policy
+  val fixed_policy : policy (* no wildcards allowed *)
+  val extensible_policy : policy (* common case *)
+  val univars_policy : policy (* fresh variables are univars (in methods) *)
+  val new_any_var : Location.t -> Env.t -> policy -> type_expr
+    (* create a new variable to represent a _; fails for fixed_policy *)
+  val new_var : ?name:string -> policy -> type_expr
+    (* create a new variable according to the given policy *)
+
+  val add_pre_univar : type_expr -> policy -> unit
+    (* remember that a variable might become a univar if it isn't unified *)
+  val collect_pre_univars : (unit -> 'a) -> 'a * type_expr list
+    (* collect univars during a computation; returns the univars
+       postcondition: the returned type_exprs are all Tunivar *)
```

Comment:
Now that you have an abstraction boundary in place, it looks like we could almost hide the concept of "pre-univar" from the user. Could we? Dummy proposal (I don't understand the full details):
- `add_pre_univar` => `add_var`
- `collect_pre_univars` => `collect_univars`

Reply:
Maybe. But I don't understand the details enough to execute here. `add_pre_univar` is called in the context of class types that I don't fully understand, so I'm hesitant to remove that.

But you're right about dropping the concept of pre-univar from the exported API. I will execute.